### PR TITLE
Fixes for TLS shutdown issues (#91)

### PR DESCRIPTION
**Pull request by hannesdejager**
This fixes the issue in #91

### CODE REVIEW

File: src/server/stream.rs [hunk 1]
```diff
@@ -248,7 +251,23 @@ impl<S: SecuritySwitch> AsyncRead for SwitchingTlsStream<S> {}
 impl<S: SecuritySwitch> AsyncWrite for SwitchingTlsStream<S> {
     fn shutdown(&mut self) -> Poll<(), io::Error> {
         debug!("AsyncWrite shutdown <<{}>>", self.channel);
-        // TODO: Anything to do for the TLS stream??
+
+        let state = self.state.lock().unwrap().which_state(self.channel);
```
Comment (wgfm) [inline src/server/stream.rs]: Is it safe to unwrap here?

File: src/server/stream.rs [hunk 2]
```diff
@@ -229,13 +229,16 @@ impl<S: SecuritySwitch> Write for SwitchingTlsStream<S> {
...
     fn flush(&mut self) -> io::Result<()> {
+        debug!("Flush called <<{}>>", self.channel);
         let state = self.state.lock().unwrap().which_state(self.channel);
         match state {
             SecurityState::On => {
                 self.tls.flush()?;
                 while self.tls.wants_write() {
                     self.tls_io(TlsIoType::Write)?;
                 }
+                self.tls.flush()?;
```
Comment (wgfm) [inline src/server/stream.rs]: Do we have to flush tls twice?

Comment (hannesdejager) [reply src/server/stream.rs]: I'm not sure to be honest and not sure how to be sure. Its precautionary.

File: src/server/stream.rs (hunk 1 above)
Comment (hannesdejager) [reply src/server/stream.rs]: the lock method will only give an error if the mutex was poisoned by a panic in another thread. In this case we probably want to panic too. Open for suggestions.

Comment (wgfm) [reply src/server/stream.rs]: I see. Just read [the Mutex documentation](https://doc.rust-lang.org/std/sync/struct.Mutex.html#poisoning). I guess we'll be fine unwrapping, then :+1:

File: src/server/stream.rs (hunk 2 above)
Comment (wgfm) [reply src/server/stream.rs]: Perhaps we don't need the first flush

Comment (hannesdejager) [reply src/server/stream.rs]: I've removed it for now.

**Pull request merged**